src-java/kilda-model/src/main/java/org/openkilda/adapter/FlowSideAdapter.java

```diff
@@ -45,41 +47,45 @@ public static FlowSideAdapter makeIngressAdapter(Flow flow, FlowPath path) {
      */
     public static FlowSideAdapter makeEgressAdapter(Flow flow, FlowPath path) {
         if (path.getCookie().getDirection() == FlowPathDirection.FORWARD) {
-            return new FlowDestAdapter(flow);
+            return new FlowDestAdapter(flow, path);
         } else {
-            return new FlowSourceAdapter(flow);
+            return new FlowSourceAdapter(flow, path);
         }
     }
 
     /**
      * Determine flow side by switchId and produce corresponding side-adapter.
      */
-    public static FlowSideAdapter makeAdapter(SwitchId switchId, Flow flow) {
+    public static FlowSideAdapter makeAdapter(SwitchId switchId, Flow flow, FlowPath flowPath) {
         if (flow.isOneSwitchFlow()) {
             throw new IllegalArgumentException("Unable to determine flow side for one-switch-flow by switch endpoint");
         } else if (switchId.equals(flow.getSrcSwitchId())) {
-            return new FlowSourceAdapter(flow);
+            return new FlowSourceAdapter(flow, flowPath);
```

Comment:
You must not make such (that this is a forward path) assumptions. At least you must make more checks.